src/system_simulation/server/simulation_runner.rb

```diff
@@ -178,25 +180,51 @@ def find_equilibrium(spring_id)
     raw_data.index(equilibrium_data_row)
   end
 
-  # This function approximates a optimum (= the biggest spring constant that makes the spring still stay in the angle
-  # constrains) by starting with a very low spring constant (which leads to a very high oscillation => high angle delta)
+  # OPTIMIZATION LOGIC
+
+  # @param Symbol kind of constrain, one of CONSTRAINS
+  # @return Hash<String, int> new constants for spring ids
+  def optimize_springs(constrain_kind)
+    # TODO: remove these mocked spring and user ids
+    # hash
+    result_map = {}
+    user_id = @mounted_users.keys[0]
+    @constants_for_springs.each do |spring_id, _constant|
+      result_map[spring_id] = optimize_spring_for_constrain(spring_id, user_id, constrain_kind)
+    end
+    result_map
+  end
+
+  # This function approximates a optimum (= the biggest spring constant that makes the spring still stay in the specific
+  # constrain) by starting with a very low spring constant (which leads to a very high oscillation => high amplitude)
   # and approaches the optimum by approaching with different step sizes (= resolutions of the search), decreasing the
   # step size as soon as the spring constant is not valid anymore and thus approximating the highest valid spring
   # constant.
-  def constant_for_constrained_angle(allowed_angle_delta = Math::PI / 2.0, spring_id = 25, initial_constant = 500)
-    # steps which the algorithm uses to approximate the valid spring constant
+  # @param [Symbol] constrain_kind specifying the kind of constrain
+  # @param [String] spring_id
+  def optimize_spring_for_constrain(spring_id, user_id, constrain_kind)
```

Comment:
What about giving this function a lambda function, instead of the ```constrain_kind```. That way we can make it reusable, and don't need data_valid_for_constrain function, which might not work when we want to have 2 seperate checks for example.

Reply:
Hmm that's a fair suggestion... I'm not too sure though: why should `data_valid_for_constrain` not work for multiple checks? If we need separate checks for a constraint we could just implement that in this function... Anyways, I would leave this for now, if we notice that we need changes when we introduce new constraints we should change it then, I guess.